master/internal/kubernetes/pods.go

```diff
@@ -108,19 +145,75 @@ func (p *pods) getMasterIPAndPort(ctx *actor.Context) error {
 	return nil
 }
 
+func (p *pods) startPodInformer(ctx *actor.Context) {
+	p.informer, _ = ctx.ActorOf("pod-informer", newInformer(p.podInterface, p.namespace, ctx.Self()))
+	ctx.Tell(p.informer, startInformer{})
+}
+
 func (p *pods) receiveStartPod(ctx *actor.Context, msg sproto.StartPod) error {
-	ref, ok := ctx.ActorOf(
-		fmt.Sprintf("pod-%s", msg.Spec.TaskID),
-		newPod(
-			p.cluster, p.clientSet, p.namespace, p.masterIP,
-			p.masterPort, msg.Spec, msg.Slots, msg.Rank,
-			p.podInterface, p.configMapInterface,
-		),
+	newPodHandler := newPod(
+		p.cluster, msg.TaskHandler, p.clientSet, p.namespace, p.masterIP,
+		p.masterPort, msg.Spec, msg.Slots, msg.Rank, p.podInterface,
+		p.configMapInterface, p.leaveKubernetesResources,
 	)
-
+	ref, ok := ctx.ActorOf(fmt.Sprintf("pod-%s-%d", msg.Spec.TaskID, msg.Rank), newPodHandler)
 	if !ok {
 		return errors.Errorf("pod actor %s already exists", ref.Address().String())
 	}
 
+	ctx.Log().WithField("pod", newPodHandler.podName).WithField(
+		"handler", ref.Address()).Infof("registering pod handler")
+
+	if _, alreadyExists := p.podNameToPodHandler[newPodHandler.podName]; alreadyExists {
+		return errors.Errorf(
+			"attempting to register same pod name: %s multiple times", newPodHandler.podName)
+	}
+
+	p.podNameToPodHandler[newPodHandler.podName] = ref
+	p.containerIDToPodHandler[msg.Spec.ContainerID] = ref
+	p.podHandlerToMetadata[ref] = podMetadata{
+		podName:     newPodHandler.podName,
+		containerID: msg.Spec.ContainerID,
+	}
+	return nil
+}
+
+func (p *pods) receivePodStatusUpdate(ctx *actor.Context, msg podStatusUpdate) {
+	ref, ok := p.podNameToPodHandler[msg.updatedPod.Name]
+	if !ok {
+		ctx.Log().WithField("pod name", msg.updatedPod.Name).Warn(
+			"received pod status update for un-registered pod")
+		return
+	}
+
+	ctx.Tell(ref, msg)
+}
+
+func (p *pods) receiveStopPod(ctx *actor.Context, msg sproto.StopPod) {
+	ref, ok := p.containerIDToPodHandler[msg.ContainerID]
+	if !ok {
+		// For multi-pod tasks, when the the chief pod exits,
+		// the scheduler will request to terminate pods all other pods
+		// that have notified the scheduler that they have exited.
+		ctx.Log().WithField("container id", msg.ContainerID).Info(
+			"received stop pod command for unregistered container id")
+		return
+	}
+
+	ctx.Tell(ref, msg)
+}
+
+func (p *pods) cleanupPodHandler(ctx *actor.Context, podHandler *actor.Ref) error {
+	podInfo, ok := p.podHandlerToMetadata[podHandler]
+	if !ok {
+		return errors.Errorf("unknown pod handler being deleted %s", podHandler)
```

Comment:
non-blocking: Logging the `actor.Ref` directly is kind of noisy -- maybe use the address instead (in `WithField`)?

Reply:
Yep updated to use the address instead. Does errors have a `WithField` ?